### ISSUE TITLE
Customise generated values based on content

### ISSUE THREAD

**Issue by rupert-madden-abbott** [instancio/instancio#213]
Sometimes, in order for generated data to be valid, constraints need to be specified across multiple fields.

For example:

```
class Foo {
  String firstName;

  String fullName;
}
```

I want to ensure that `fullName` is prefixed with `firstName`;

I would like to be able to write:

```
Instancio.of(Foo.class)
  .generate(field("fullName"), gen -> foo -> gen.string().prefix(foo.getFirstName() + " ")
  .create();
```

I would like fields with specs specified in this way to be applied after other fields have been generated and in the order specified, so that fields that do not have dependencies can be generated first, and then depended on by later fields.

**Comment by armandino** [comment]
Thanks for the suggestion. Have you considered using an `onComplete()` callback to handle this case?
Assuming that `Foo` is mutable, it should be possible to update the field after the object has been populated:

```java
Instancio.of(Foo.class)
    .onComplete(all(Foo.class), (Foo foo) -> {
        foo.setFullName(foo.getFirstName() + " " + foo.getLastName()));
    })
    .create();
```

If the class is immutable, then it can still be done, but a little more verbose:

```java
String firstName = "foo";
String lastName = "bar";
Instancio.of(Foo.class)
        .set(field("firstName"), firstName)
        .set(field("lastName"), lastName)
        .set(field("fullName"), firstName + " " + lastName))
        .create();
```

As for having fields generated in a specific order, that is not something that can be done without significant changes to the library, although it would be a nice feature.

**Comment by rupert-madden-abbott** [comment]
No problem, thanks for the library.

I had considered `onComplete` but the problem is twofold:

* I do want to use this with immutable objects
* I also want to use this with [globally configured patterns](https://github.com/instancio/instancio/issues/212). For example, lets say I have an indefinite number of types, all of which have Instant fields "createdOn" and an "updatedOn" and I want to ensure that updatedOn is later than createdOn. I don't want to have to call "onComplete" for each possible type (because I don't know all of the specific types in my organisation specific test library).

Pre-generating the data is an interesting idea but suffers from the same pitfalls (and means I can't take advantage of the random generation. For example, I'm no longer getting random first names with this approach. If my depended on data were part of a unique key, I then need to develop a strategy outside of this library to try and avoid conflicts.

I completely appreciate that implementing this kind of ordering may be complex (and possibly error prone).

How about allowing the seed to be shared between multiple fields such that they would then predictably produce the same values?

Would something like this work (although it feels a bit of an unintended usage of the API):

```
Instancio.of(Foo.class)
  .supply(field("firstName"), () -> Instancio.of(String.class).withSeed(1234).create())
  .generate(field("fullName"), gen -> gen.string().prefix(Instancio.of(String.class).withSeed(1234).create() + " ")
  .create();
```

**Comment by armandino** [comment]
Thanks for the additional context. I think I understand your goal better now. The last approach in your latest comment should work. How it's actually implemented will depend on whether you use `InstancioExtension` (requires JUnit 5).

With the extension it should be simple, because the extension [manages the seed](https://www.instancio.org/user-guide/#seed-lifecycle-in-a-junit-jupiter-test) for us. Therefore, we can make different calls to `Instancio.create(...)` within a `@Test` method without specifying the seed. Each `@Test` method is isolated and objects within it are automatically created based on the same seed value. This ensure the data is reproducible.

Without `InstancioExtension`, it should also be possible, but in this case we will need to set the seed manually on each call to `Instancio.create()`:

```java
Instancio.of(...).withSeed(123);

// or using an instance of Settings
Settings settings = Settings.create().set(Keys.SEED, 123L);
Instancio.of(...).withSettings(settings);
```

One option to keep this manageable would be to create a factory class for creating objects and maintain the `seed` as an instance variable, so it can be shared across different methods.

Regarding re-use of generation parameters when creating objects of different types, it should be possible using Models. For example, we can define a common model for classes that have `lastModified` and `createdOn` fields:

```java
static <T> Model<T> commonModel(Class<T> klass) {
    LocalDateTime lastModified = Instancio.of(LocalDateTime.class)
            .generate(all(LocalDateTime.class), gen -> gen.temporal().localDateTime().past())
            .create();

    return Instancio.of(klass)
            .set(field("lastModified"), lastModified)
            .generate(field("createdOn"), gen -> gen.temporal().localDateTime().range(LocalDateTime.MIN, lastModified))
            .toModel();
}
```

Now it should be possible to generate an instance of any class that has these fields using the model:

```java
Model<Foo> fooModel = commonModel(Foo.class);
Foo foo = Instancio.create(fooModel);
assertThat(foo.getLastModified()).isAfterOrEqualTo(foo.getCreatedOn());
```

I created a new repo with a sample test class [here](https://github.com/instancio/instancio-github-issues). Hope this works as a workaround.

**Comment by bgulowaty** [comment]
> Thanks for the additional context. I think I understand your goal better now. The last approach in your latest comment should work. How it's actually implemented will depend on whether you use `InstancioExtension` (requires JUnit 5).
> 
> With the extension it should be simple, because the extension [manages the seed](https://www.instancio.org/user-guide/#seed-lifecycle-in-a-junit-jupiter-test) for us. Therefore, we can make different calls to `Instancio.create(...)` within a `@Test` method without specifying the seed. Each `@Test` method is isolated and objects within it are automatically created based on the same seed value. This ensure the data is reproducible.
> 
> Without `InstancioExtension`, it should also be possible, but in this case we will need to set the seed manually on each call to `Instancio.create()`:
> 
> ```java
> Instancio.of(...).withSeed(123);
> 
> // or using an instance of Settings
> Settings settings = Settings.create().set(Keys.SEED, 123L);
> Instancio.of(...).withSettings(settings);
> ```
> 
> One option to keep this manageable would be to create a factory class for creating objects and maintain the `seed` as an instance variable, so it can be shared across different methods.
> 
> Regarding re-use of generation parameters when creating objects of different types, it should be possible using Models. For example, we can define a common model for classes that have `lastModified` and `createdOn` fields:
> 
> ```java
> static <T> Model<T> commonModel(Class<T> klass) {
>     LocalDateTime lastModified = Instancio.of(LocalDateTime.class)
>             .generate(all(LocalDateTime.class), gen -> gen.temporal().localDateTime().past())
>             .create();
> 
>     return Instancio.of(klass)
>             .set(field("lastModified"), lastModified)
>             .generate(field("createdOn"), gen -> gen.temporal().localDateTime().range(LocalDateTime.MIN, lastModified))
>             .toModel();
> }
> ```
> 
> Now it should be possible to generate an instance of any class that has these fields using the model:
> 
> ```java
> Model<Foo> fooModel = commonModel(Foo.class);
> Foo foo = Instancio.create(fooModel);
> assertThat(foo.getLastModified()).isAfterOrEqualTo(foo.getCreatedOn());
> ```
> 
> I created a new repo with a sample test class [here](https://github.com/instancio/instancio-github-issues). Hope this works as a workaround.

This way makes the local date fixed. The perfect solution would be, if - during the generation of new object - other generator/supplier could access already generated "state" dynamically through some kind of context. I haven't used it, but by looking at the docs, EasyRandom's `ContextAwareRandomizer` implements such approach. 

I think it would greatly improve usability of the lib. 
Use case I've had, while stumbling at this issue is:

> I want to generate object/model, where one of two string fields can be missing. Object can either have "name" or "companyName" passed through constructor and at least one of those fields need to be set, while other one can be missing.

**Comment by armandino** [comment]
@bgulowaty thank you for the feedback! I agree with you that it would be a nice feature to support this.

I did a quick test with `ContextAwareRandomizer`, however I'm not convinced that it's the right approach as it has a couple of drawbacks:

1. The result depends on the class structure.
2. The result depends on the internal implementation details (i.e. the order in which fields are populated).

For example, given this class:

```java
class Person {
    String email;
    String country;
}
```

you'd like to generate:

- `xyz@mail.se` if the country is Sweden
- `xyz@mail.no`if the country is Norway

If `email` is generated first, then this approach doesn't really work because the value of `country` is not yet known.
However, if you re-arrange the fields and place `country` first, then it works. I think it would be a bad user experience if their tests suddenly started to fail because somebody simply changed the order of the fields.

In addition, if the library's implementation changes the order in which it processes the fields, it would also break existing tests.

I'll continue exploring the options to support this type of feature, but right now I don't have a solution.

**Comment by armandino** [comment]
I just released version `3.0.0` which adds support for the use cases described in this issue using the new `assign()` API.

Please check out the [release notes](https://github.com/orgs/instancio/discussions/687) and the [user guide](https://www.instancio.org/user-guide/#using-assign)
